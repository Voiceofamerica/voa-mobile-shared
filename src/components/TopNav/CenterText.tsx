
import * as React from 'react'

import { centerText } from './TopNav.scss'

export interface Props extends React.Props<any> {
}

function CenterText (props: Props) {
  const { children } = props

  return (
    <div className={centerText}>
      {children}
    </div>
  )
}

export default CenterText
