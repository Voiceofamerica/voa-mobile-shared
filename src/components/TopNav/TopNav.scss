
@import '../../commonStyles/index.scss';

.topNav {
  position: fixed;
  display: flex;
  flex-direction: row;
  align-content: stretch;
  flex-wrap: nowrap;
  z-index: $menu-z;
  top: 0;
  left: 0;
  right: 0;
  background: white;
  box-shadow: 0 2px 2px rgba(0, 0, 0, 0.5);
  height: $menu-height;
  overflow-x: scroll;
  font-size: $menu-font-size;

  &::-webkit-scrollbar {
    display: none;
  }

  .topNavItem {
    @include touchable();

    flex-grow: 1;
    flex-wrap: nowrap;
    display: flex;
    flex-direction: row;
    align-items: center;
    padding: 0 10px;
    white-space: nowrap;
    background: transparent;
    outline: none;
    border: none;
    border-bottom: 3px solid transparent;
    border-radius: 0;
    font-size: 1em;
    color: #888;
    box-sizing: border-box;

    &.selectedItem {
      border-bottom-color: $voa-red;
      color: $voa-blue;
      font-weight: bold;
    }

    &:last-of-type {
      min-width: 100vw;
    }
  }
  
  .fadeout {
    position: fixed;
    pointer-events: none;
    background: linear-gradient(90deg, rgba(256, 256, 256, 0), #fff 90%, #fff);
    top: 0;
    right: 0;
    height: $menu-height;
    width: 30vw;
    z-index: $menu-z + 1;

    [dir="rtl"] & {
      right: auto;
      left: 0;
      background: linear-gradient(-90deg, rgba(256, 256, 256, 0), #fff 90%, #fff);
    }
  }

  .centerText {
    display: flex;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    align-items: center;
    font-weight: bold;
    justify-content: center;
    color: $voa-blue;
    z-index: $menu-z + 2;
  }
}
