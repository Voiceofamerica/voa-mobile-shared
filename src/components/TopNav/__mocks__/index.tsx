
import * as React from 'react'

export default (props) => (
  React.createElement('top-nav', { props }, props.children)
)

export const TopNavItem = (props) => (
  React.createElement('top-nav-item', { props }, props.children)
)

export const CenterText = (props) => (
  React.createElement('center-text', { props }, props.children)
)

export const StaticItem = (props) => (
  React.createElement('static-item', { props }, props.children)
)
