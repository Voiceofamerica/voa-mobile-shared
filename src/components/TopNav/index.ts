
export { default } from './TopNav'
export { default as TopNavItem } from './TopNavItem'
export { default as CenterText } from './CenterText'
export { default as StaticItem } from './StaticItem'
