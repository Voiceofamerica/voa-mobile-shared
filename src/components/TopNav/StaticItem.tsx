
import * as React from 'react'

import { staticItem } from './TopNav.scss'

export interface Props extends React.Props<any> {
  style?: React.CSSProperties
  onClick?: () => void
}

function StaticItem (props: Props) {
  const { children, onClick = () => null, style } = props

  return (
    <button className={staticItem} onClick={() => onClick()} style={style}>
      {children}
    </button>
  )
}

export default StaticItem
