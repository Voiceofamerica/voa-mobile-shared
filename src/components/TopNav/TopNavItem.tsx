
import * as React from 'react'

import { topNavItem, selectedItem } from './TopNav.scss'

export interface Props extends React.Props<any> {
  selected?: boolean
  onClick?: () => void
}

function TopNavItem (props: Props) {
  const { children, selected, onClick = () => null } = props

  const className = selected
                  ? `${topNavItem} ${selectedItem}`
                  : topNavItem

  return (
    <button className={className} onClick={() => onClick()}>
      {children}
    </button>
  )
}

export default TopNavItem
