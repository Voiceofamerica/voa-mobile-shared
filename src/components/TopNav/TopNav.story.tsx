
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import ThemeProvider, { BaseTheme, TopNavTheme } from '../ThemeProvider'

import TopNav from './TopNav'
import CenterText from './CenterText'
import StaticItem from './StaticItem'
import TopNavItem from './TopNavItem'

const itemNames = [
  'Headlines',
  'US Politics',
  'USA',
  'Sports',
  'Cars',
  'Technology',
  'Foreign Policy',
  'Foreign Politics',
  'Weather',
  'Food',
  'The Economy',
  'China Relations',
  'Russia Relations',
]

interface SelectableTopNavProps {
  prefix?: string
  suffix?: string
  rtl?: boolean
}

interface SelectableTopNavItemsState {
  selectedIndex: number
}

const testBaseTheme: BaseTheme = {
  primaryColor: '#35A527',
  grey: '#666666',
  lightGrey: '#AAAAAA',
  white: '#CCFFCC',
  darkGrey: '#333366',
}

const testLocalTheme: TopNavTheme = {
  topNavBackground: '#ffaabb',
  topNavStaticColor: '#993246',
  topNavColor: '#333333',
  topNavSelectedBackground: '#3527A5',
  topNavSelectedColor: '#EEEEEE',
}

class SelectableTopNavItems extends React.Component<SelectableTopNavProps, SelectableTopNavItemsState> {
  state: SelectableTopNavItemsState = {
    selectedIndex: 0,
  }

  render () {
    const { prefix, suffix, rtl } = this.props
    const { selectedIndex } = this.state

    return (
      <TopNav rtl={rtl}>
        { prefix && <StaticItem onClick={action('prefix clicked')}>{prefix}</StaticItem> }
        {
          itemNames.map((name, idx) => (
              <TopNavItem key={idx} selected={idx === selectedIndex} onClick={this.onClick(idx)}>
                {name}
              </TopNavItem>
          ))
        }
        { suffix && <StaticItem onClick={action('suffix clicked')}>{suffix}</StaticItem> }
      </TopNav>
    )
  }

  private onClick = (selectedIndex: number) => () => {
    this.setState({ selectedIndex })
  }
}

storiesOf(TopNav.name, module)
  .add('display', () => (
    <SelectableTopNavItems />
  ))
  .add('display with static prefix', () => (
    <SelectableTopNavItems prefix='A' />
  ))
  .add('display with static suffix', () => (
    <SelectableTopNavItems suffix='Z' />
  ))
  .add('display with both static elements', () => (
    <SelectableTopNavItems prefix='A' suffix='Z' />
  ))
  .add('base themed display', () => (
    <ThemeProvider value={testBaseTheme}>
      <SelectableTopNavItems prefix='A' suffix='Z' />
    </ThemeProvider>
  ))
  .add('local themed display', () => (
    <ThemeProvider value={testLocalTheme}>
      <SelectableTopNavItems prefix='A' suffix='Z' />
    </ThemeProvider>
  ))
  .add('rtl display', () => (
    <div dir='rtl'>
      <SelectableTopNavItems rtl />
    </div>
  ))
  .add('rtl display with both static elements', () => (
    <div dir='rtl'>
      <SelectableTopNavItems prefix='A' suffix='Z' rtl />
    </div>
  ))
