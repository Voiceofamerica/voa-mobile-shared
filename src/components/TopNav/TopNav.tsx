
import * as React from 'react'

import StaticItem from './StaticItem'
import TopNavItem, { Props as TopNavItemProps } from './TopNavItem'
import CenterText from './CenterText'
import { topNav, innerContainer, itemContainer, fadeout } from './TopNav.scss'

export interface ColorInfo {
  mainBackground: string
  mainColor: string
  staticColor: string
  selectedBackground: string
  selectedColor: string
}

export interface Props extends React.Props<any> {
  color?: ColorInfo
  rtl?: boolean
}

const DEFAULT_COLORS: ColorInfo = {
  mainBackground: '#EEEEEE',
  mainColor: '#333333',
  staticColor: '#B4BCC2',
  selectedBackground: '#D41010',
  selectedColor: '#FFFFFF',
}

function toRGB (color: string) {
  const r = parseInt(color.substring(1, 3), 16)
  const g = parseInt(color.substring(3, 5), 16)
  const b = parseInt(color.substring(5, 7), 16)

  return `${r}, ${g}, ${b}`
}

function TopNav ({ children, color = DEFAULT_COLORS, rtl }: Props) {
  const {
    mainBackground,
    mainColor,
    staticColor,
    selectedBackground,
    selectedColor,
  } = color

  const rChildren = React.Children.toArray(children) as React.ReactElement<any>[]
  const items = rChildren.filter(({ type }) => type === TopNavItem || type === CenterText)
                         .map(child => {
                           if (child.type === CenterText) {
                             return child
                           } else {
                             const props = child.props as TopNavItemProps
                             const { key } = child
                             console.log('props', props)

                             if (!props.selected) {
                               return child
                             } else {
                               console.log('got here')
                               return <TopNavItem key={key} {...props} style={{ ...props.style, color: selectedColor, background: selectedBackground }} />
                             }
                           }
                         })
  const statics = rChildren.filter(child => child.type === StaticItem)
                           .map(({ key, props }) => (
                             <StaticItem key={key} {...props} style={{ ...props.style, color: staticColor }} />
                           ))

  const rgb = toRGB(mainBackground)
  const fadeoutBackground = rtl
                          ? `linear-gradient(-90deg, transparent, rgba(${rgb}, 0.9) 90%)`
                          : `linear-gradient(90deg, transparent, rgba(${rgb}, 0.9) 90%)`
  return (
    <div className={topNav} style={{ background: mainBackground, color: mainColor }}>
      {statics[0]}
      <div className={itemContainer}>
        <div className={innerContainer}>
          {items}
        </div>
        <div className={fadeout} style={{ backgroundImage: fadeoutBackground }} />
      </div>
      {statics[1]}
    </div>
  )
}

export default TopNav
