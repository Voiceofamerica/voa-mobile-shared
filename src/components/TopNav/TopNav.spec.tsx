
import * as React from 'react'
import { create } from 'react-test-renderer'

import TopNav from './TopNav'

jest.mock('../ThemeProvider')

describe(`<${TopNav.name} />`, () => {
  describe('shapshots', () => {
    it('should render empty', () => {
      const element = create((
        <TopNav />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given children', () => {
      const element = create((
        <TopNav>
          <div>child1</div>
          <div>child2</div>
        </TopNav>
      ))

      expect(element).toMatchSnapshot()
    })
  })
})
