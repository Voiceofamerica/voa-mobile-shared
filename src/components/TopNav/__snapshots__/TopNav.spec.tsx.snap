// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopNav /> shapshots should render empty 1`] = `
<div
  className="topNav"
  style={
    Object {
      "background": "#000000",
      "color": "#000000",
    }
  }
>
  <div
    className="itemContainer"
  >
    <div
      className="innerContainer"
    >
      <div
        className="topNavItem"
      >
         
      </div>
    </div>
    <div
      className="fadeout"
      style={
        Object {
          "backgroundImage": "linear-gradient(90deg, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.9) 90%)",
        }
      }
    />
  </div>
</div>
`;

exports[`<TopNav /> shapshots should render with the given children 1`] = `
<div
  className="topNav"
  style={
    Object {
      "background": "#000000",
      "color": "#000000",
    }
  }
>
  <div
    className="itemContainer"
  >
    <div
      className="innerContainer"
    >
      <div
        className="topNavItem"
      >
         
      </div>
    </div>
    <div
      className="fadeout"
      style={
        Object {
          "backgroundImage": "linear-gradient(90deg, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.9) 90%)",
        }
      }
    />
  </div>
</div>
`;
